pageserver/src/tenant/storage_layer.rs

```diff
@@ -140,8 +142,17 @@ impl ValuesReconstructState {
         }
     }
 
-    pub(crate) fn on_layer_visited(&mut self) {
+    pub(crate) fn on_layer_visited(&mut self, layer: &ReadableLayer) {
         self.layers_visited += 1;
+        if let ReadableLayer::PersistentLayer(layer) = layer {
+            if layer.layer_desc().is_delta() {
+                self.delta_layers_visited += 1;
+            }
+        }
+    }
+
+    pub(crate) fn get_delta_layers_visited(&self) -> u32 {
```

Comment:
Once you merge https://github.com/neondatabase/neon/pull/7773, `get_layers_visited` will become a very good approximation for this. Since you're not splitting image layers, it will basically return the number of delta layers plus one. Hence, I don't think you need this and the associated change in `on_layers_visited`.

Reply:
but it's still an approximation, and once someone modifies the image layer generation logic, they will likely spend some time fixing existing test cases relying on that assumption...? So I'm in favor of adding this statistics, and as you said, in a more light-weight way inside reconstruct value state.

Reply:
> but it's still an approximation, and once someone modifies the image layer generation logic, they will likely spend some time fixing existing test cases relying on that assumption...?

Not sure what you mean.

> in a more light-weight way inside reconstruct value state

+1

Reply:
... also need to consider if we generate multiple image layers for the metadata keyspace (which I think will happen very soon), so I think it's good to have these two separate metrics on layer visited for now.

Reply:
> also need to consider if we generate multiple image layers for the metadata keyspace

That will break the scan interface with the changes in #7773.

In any case `layers_visited` would be off by at most the number of image layers, but if you feel strongly about it, then let's keep it